pip/pip-425.md

```diff
@@ -0,0 +1,147 @@
+
+
+# Goals
+Optimize the PulsarClient code logic, remove unavailable serviceUrl endpoints through the feedback mechanism, and improve the success rate of PulsarClient requests.
+
+## In Scope
+- Improve Connection Resilience: Modify the PulsarClient to avoid repeatedly trying to connect to unavailable endpoints.
+- Introduce Endpoint Quarantining: When a connection to an endpoint fails, temporarily remove it from the pool of available endpoints.
+- Implement Exponential Backoff: Re-introduce quarantined endpoints for connection attempts after a dynamically increasing delay, allowing them to recover without overwhelming the system.
+
+# Detailed Design
+We will introduce an endpoint quarantining strategy based on passive feedback from connection attempts.
+**Explanation of nouns:**
+- `availableAddressPool`: a pool containing all available addresses, subset of serviceUrl.
+- `allAddressPool`: a pool containing all addresses in serviceUrl.
+
+By default, the client will only try to connect to endpoints from the availableAddressPool.
+
+- On Connection Failure: If the client fails to connect to an endpoint, it will be temporarily removed from the availableAddressPool and placed into "quarantine".
+- Exponential Backoff: The quarantine duration will increase exponentially upon consecutive failures.
+- On Connection Success: A successful connection to an endpoint proves it is available. If the endpoint was previously failing, its quarantine backoff state will be reset. It will be immediately considered part of the availableAddressPool.
+- Fallback: If the availableAddressPool ever becomes empty, the client will fall back to trying a random endpoint from the allAddressPool. This prevents a deadlock where the client cannot connect because all known endpoints are simultaneously in quarantine.
+As [#22934](https://github.com/apache/pulsar/issues/22934#issuecomment-2943690398) mentioned, add `markHostAvailability` method in `ServiceNameResolver` interface like:
+```java
+    /**
+     * Mark the availability of a host.
+     * @param address the host address to mark availability for
+     * @param isAvailable true if the host is available, false otherwise
+     */
+    default void markHostAvailability(InetSocketAddress address, boolean isAvailable){
+        // Default implementation does nothing
+        // Subclass can override this method to implement host availability tracking
+    }
+```
+When resolve one address in serviceUrl, the `ConnectionPool` will invoke `markHostAvailability` to tell availability about that address, so the `ServiceNameResolver` can remove it from `availableAddressPool` if unavailable (or recover back to `availableAddressPool` if available). If all addressed in  `availableAddressPool` are unavailable, the resolver implementation could start getting random addresses from the `allAddressPool`.
```

Comment:
It seems we need to add a circuit breaker mechanism. Can we simply address your problem by:
1. Changing from establishing connections by randomly selecting nodes to using a round-robin approach.
2. To prevent hotspots caused by multiple clients connecting to the same node , shuffling the list of IPs used for round-robin.

Reply:
OK, that was a typo, what I meant was to use a round-robin approach to select endpoint.

Reply:
https://github.com/apache/pulsar/blob/5841238d49695e29f557e22ce964d5b7470a354a/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarServiceNameResolver.java#L46-L58

@315157973 OKay, maybe I didn't explain it clearly enough.
This pip will not change the algorithm for selecting address, and will still maintain the original round-robin algorithm. What we change is the value of the list variable in the `resolveHost` function. When there are available addresses, a address are selected from the available address list. When there are no available addresses,  fall back to the original list (all address passed in by serviceUrl) for selection.